src/operators/products.c

```diff
@@ -0,0 +1,221 @@
+   IGRAPH_SAFE_MULT(vcountg1, ecountg2, &ecount);
+   IGRAPH_SAFE_MULT(vcountg2, ecountg1, &temp);
+   IGRAPH_SAFE_ADD(ecount, temp, &ecount);
+   IGRAPH_SAFE_MULT(ecount, 2, &ecount);
+   IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, ecount);
+
+   igraph_integer_t from, to;
+   igraph_integer_t i, j;
+   
+   // Vertex ((i, j)) with i from g1, and j from g2
+   //   will have new vertex id: i * vcountg2 + j
+
+   igraph_integer_t edge_index = 0;
+   // Edges from g1
+   for (i = 0; i < ecountg1; ++i) {
+      from = IGRAPH_FROM(g1, i);
+      to = IGRAPH_TO(g1, i);
+      // for all edges (from, to) in g1, add edge from ((from, j)) to ((to, j))
+      //    for all vertex j in g2
+      for (j = 0; j < vcountg2; ++j) {
+         // SAFE MULT and SAFE ADD not needed as < vcount
+         VECTOR(edges)[edge_index++] = from * vcountg2 + j;   // ((from, j))
+         VECTOR(edges)[edge_index++] = to * vcountg2 + j;     // ((to, j))
+      }
+   }
+
+   // Edges from g2
+   for (i = 0; i < ecountg2; ++i) {
+      from = IGRAPH_FROM(g2, i);
+      to = IGRAPH_TO(g2, i);
+      // for all edges (from, to) in g2, add edge from (j, from) to (j, to)
+      //    for all vertex j in g1
+      for (j = 0; j < vcountg1; ++j) {
+         VECTOR(edges)[edge_index++] = j * vcountg2 + from; // ((j, from))
+         VECTOR(edges)[edge_index++] = j * vcountg2 + to;   // ((j, to))
+      }
+   }
+
+   IGRAPH_CHECK(igraph_create(res, &edges, vcount, directed));
+   igraph_vector_int_destroy(&edges);
+   IGRAPH_FINALLY_CLEAN(1);
+
+   return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t tensor_product(igraph_t *res,
+   const igraph_t *g1, const igraph_t *g2) {
+      
+   igraph_bool_t directed1 = igraph_is_directed(g1);
+   igraph_bool_t directed2 = igraph_is_directed(g2);
+
+   if (directed1 != directed2) {
+      IGRAPH_ERROR("Tensor product between a directed and an undirected graph is invalid. "
+         "Please ensure that both graphs have the same directionality.", IGRAPH_EINVAL);
+   }
+
+   igraph_bool_t directed = directed1 && directed2;
+
+   igraph_integer_t vcountg1 = igraph_vcount(g1);
+   igraph_integer_t vcountg2 = igraph_vcount(g2);
+   igraph_integer_t ecountg1 = igraph_ecount(g1);
+   igraph_integer_t ecountg2 = igraph_ecount(g2);
+   igraph_integer_t vcount;
+   igraph_integer_t ecount;
+
+   IGRAPH_SAFE_MULT(vcountg1, vcountg2, &vcount);
+   igraph_vector_int_t edges;
+
+   // new edge count = 2*e1*e2 if undirected else e1*e2
+   IGRAPH_SAFE_MULT(ecountg1, ecountg2, &ecount);
+   if (!directed) { // directed tensor product has only e1*e2 edges, see below
+      IGRAPH_SAFE_MULT(ecount, 2, &ecount);
+   }
+   IGRAPH_SAFE_MULT(ecount, 2, &ecount);
+   IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, ecount);
+
+   igraph_integer_t from1, to1, from2, to2;
+   igraph_integer_t i, j;
+   
+   // Vertex ((i, j)) with i from g1, and j from g2
+   //   will have new vertex id: i * vcountg2 + j
+   igraph_integer_t edge_index = 0;
+
+   for (i = 0; i < ecountg1; ++i) {
```

Comment:
Let's aim to declare iterator variables within `for (...` directly. Making variable scope as local as possible helps readability.

It is only in rare cases that this is not possible or not beneficial to readability.